File: Study/linkedList.py

class Node:
  def __init__(self, value, next = None):
    self.value = value
    self.next = next
    

class LinkedList:
  def __init__(self):
    self.head = None
    self.tail = None

  def append(self, value):
    node = Node(value)

    if self.head is None:
      self.head = node
      self.tail = node
    else:
      self.tail.next = node
      self.tail = node

  def removeLast(self):
    if self.head is None:
      return
    
    if self.head.next is None:
      self.head = None
      self.tail = None
      return
    
    cur = self.head
    while cur.next.next is not None:
      cur = cur.next

    cur.next = None
    self.tail = cur

  def remove(self, index):
    if self.head is None:
      return
    
    if index == 0:
      self.head = self.head.next
      if self.head is None:
        self.tail = None
      return

    cur = self.head
    for _ in range(index-1):
      if cur.next.next is None:
        return
      cur = cur.next
  
    tmp = cur.next
    cur.next = cur.next.next
    
    if self.tail == tmp:
      self.tail = cur

  def print(self):
    if self.head is None:
      print("first and last is None")
      return
    
    print("first:", self.head.value)
    print("last:", self.tail.value)

    cur = self.head
    while cur is not None:
      print(cur.value)
      cur = cur.next


linkedList = LinkedList()

linkedList.append(1)
linkedList.append(2)
linkedList.append(3)
linkedList.remove(1)
linkedList.remove(2)
linkedList.print()
linkedList.remove(0)

linkedList.print()

File: Swift-Study/LinkedList.swift
//
//  LinkedList.swift
//  Swift-Study
//
//  Created by 이정동 on 1/3/25.
//

import Foundation

class LinkedList<T> {
  class Node {
    var next: Node?
    var value: T
    
    init(next: Node? = nil, value: T) {
      self.next = next
      self.value = value
    }
  }
  
  var head: Node?
  var tail: Node?
  
  func append(value: T) {
    let node = Node(value: value)
    
    if head == nil {
      self.head = node
      self.tail = node
      return
    }
    
    self.tail?.next = node
    self.tail = node
  }
  
  func removeLast() {
    if self.head == nil {
      return
    }
    
    if self.head?.next == nil {
      self.head = nil
      self.tail = nil
      return
    }
    
    var cur = head
    while cur?.next?.next != nil {
      cur = cur?.next
    }
    
    cur?.next = nil
    self.tail = cur
  }
  
  func remove(at index: Int) {
    if self.head == nil {
      return
    }
    
    if index == 0 {
      self.head = self.head?.next
      if self.head == nil {
        self.tail = nil
      }
      return
    }
    
    var cur = self.head
    for _ in 0..<index-1 {
      if cur?.next?.next == nil {
        return
      }
      cur = cur?.next
    }
    
    cur?.next = nil
    self.tail = cur
  }
  
  func print() {
    Swift.print(self.head?.value, self.tail?.value)
  }
}
